Improve contributors documentation on running tests locally (#2167)
cocotb/cocotb: CONTRIBUTING.md

This guide explains how to contribute to cocotb, and documents the processes we agreed on to manage the project.
All processes in this document are designed to streamline the development effort, to avoid bottlenecks, and to ultimately give a pleasant experience to all involved.

## Setting Up a Development Environment

Assuming you have used cocotb prior to reading this guide, you will already have the cocotb [installation prerequisites](https://docs.cocotb.org/en/latest/install.html) and standard development tools (editor, shell, git, etc.) installed.

Additionally, you will need [doxygen](https://www.doxygen.nl/index.html), for building documentation, and [tox](https://pypi.org/project/tox/), for building documentation and running regression tests.

We recommend if you are using a Linux distribution to use your system package manager to install doxygen.
Likewise, doxygen can be installed using the homebrew package manager on Mac OS.
Windows contributors should download a binary distribution installer from the main website.

> `tox` is a Python project and can be installed with `pip`.

Reviewer: ```suggestion
`tox` is a Python project and can be installed with `pip`:
```

```command
pip install tox
```

Finally, you should [fork and clone](https://guides.github.com/activities/forking/) the cocotb repo.

> This will allows you to make changes to cocotb source code, and run regressions and build documentation locally.

Reviewer: ```suggestion
This will allow you to make changes to cocotb the source code, and run regressions and build documentation locally.
```

Author: ```suggestion
This will allow you to make changes to the cocotb source code, and run regressions and build documentation locally.
```
I think you put the `the` in the wrong place.

Now you are ready to contribute!

## Running Tests Locally

First, [set up your development environment](#setting-up-a-development-environment).

Our tests are managed by `tox`, which runs both `pytest` tests and our system of makefiles.
The regression does not end on the first failure, but continues until all tests in the `/tests` and `/examples` directories have been run.

To run the tests locally with `tox`, you will need to select an appropriate test environment.
Valid test environments are formatted as `{your python version}-{your OS}`.

> Valid python version values are `py35`, `py36`, `py37`, `py38`, or `py39`;

Reviewer: ```suggestion
Valid python version values are e.g. `py36`, or `py39`;
```
Maybe formulate this as examples, not an exhaustive list, so that this is applicable in the future (``py310``) without changes.

Author: It would be nice to be able to point to what the list is though, which is the 3rd code section [here](https://tox.readthedocs.io/en/latest/example/basic.html#a-simple-tox-ini-default-environments). Put I cannot get a direct link to it.

Author: Those are the ones we list on the tox envlist as valid. It is intended to update that after the 1.5 release to remove 3.5 and add 3.10 to show which Python versions are supported.

Reviewer: Then let's keep the full list and add a comment to the ``tox.ini``.

and valid OS values are `linux`, `macos`, or `windows`.
For example, a valid test environment is `py38-linux`.

> You can see the list of valid test environments by running the below command.

Reviewer: ```suggestion
You can see the list of valid test environments by running the below command:
```

```command
tox -l
```

> Once you know the test environment you wish to use, call `tox` .

Reviewer: ```suggestion
Once you know the test environment you wish to use, call `tox`.
```

```command
tox -e py38-linux
```

At the end of the regression, if there were any test failures, the tests that failed will be printed.
Otherwise, tox will print a green `:)`.

### Selecting a Language and Simulator for Regression

> `tox` supports the usage of the environment variables [`SIM`](https://docs.cocotb.org/en/stable/building.html#var-SIM) and [`TOPLEVEL_LANG`](https://docs.cocotb.org/en/stable/building.html#var-TOPLEVEL_LANG) to select a simulator and language to run the regression.

Reviewer: ```suggestion
`tox` supports the usage of the environment variables [`SIM`](https://docs.cocotb.org/en/latest/building.html#var-SIM) and [`TOPLEVEL_LANG`](https://docs.cocotb.org/en/latest/building.html#var-TOPLEVEL_LANG) to select a simulator and language to run the regression.
```
I'd use ``latest`` docs since this is about developing.

By default the tests will attempt to run with the Icarus Verilog simulator.

> For example, if you wanted to run tests with GHDL on Linux with Python 3.8, you would issue the following command.

Reviewer: ```suggestion
For example, if you wanted to run tests with GHDL on Linux with Python 3.8, you would issue the following command:
```
?

```command
SIM=ghdl TOPLEVEL_LANG=vhdl tox -e py38-linux
```

### Running Individual Tests Locally

Each test under `/tests/test_cases/*/` and `/examples/*/tests/` can be run individually.
This is particularly useful if you want to run a particular test that fails the regression.

> First you must install cocotb from source by navigating to the project root directory and issuing the following.

Reviewer: ```suggestion
First you must install cocotb from source by navigating to the project root directory and issuing the following command:
```

```command
python -m pip install .
```

> On Windows, you must instead install cocotb from source like so.

Reviewer: ```suggestion
On Windows, you must instead install cocotb from source like so:
```

```command
python -m pip install --global-option build_ext --global-option --compiler=mingw32 .
```

Once that has been done, you can navigate to the directory containing the test you wish to run.

> Then you may issue an [appropriate]https://docs.cocotb.org/en/stable/building.html#makefile-based-test-scripts) `make` command.

Reviewer: ```suggestion
Then you may issue an [appropriate]https://docs.cocotb.org/en/stable/building.html#makefile-based-test-scripts) `make` command:
```

```command
make SIM=icarus
```

## Building Documentation Locally

First, [set up your development environment](#setting-up-a-development-environment).

Documentation is built locally using `tox`.

> The last message in the output will contain a URL to the documentation you just built.

Reviewer: I'd move this part a bit more down, after the exact command has been shown.

Simply copy and paste the link into your browser to view it.
The documentation will be built in the same location on your hard drive on every run, so you only have to refresh the page to see new changes.

> To build the documentation locally on Linux or Mac, issue the following command.

Reviewer: ```suggestion
To build the documentation locally on Linux or Mac, issue the following command:
```

All processes in this document are designed to streamline the development effort, to avoid bottlenecks, and to ultimately give a pleasant experience to all involved.

## Setting Up a Development Environment

Assuming you have used cocotb prior to reading this guide, you will already have the cocotb [installation prerequisites](https://docs.cocotb.org/en/latest/install.html) and standard development tools (editor, shell, git, etc.) installed.

Additionally, you will need [doxygen](https://www.doxygen.nl/index.html), for building documentation, and [tox](https://pypi.org/project/tox/), for building documentation and running regression tests.

We recommend if you are using a Linux distribution to use your system package manager to install doxygen.
Likewise, doxygen can be installed using the homebrew package manager on Mac OS.
Windows contributors should download a binary distribution installer from the main website.

`tox` is a Python project and can be installed with `pip`.

```command
pip install tox
```

Finally, you must download the cocotb source from Github if you have not already done so.

```command
git clone https://github.com/cocotb/cocotb

> ```

Author: Change to fork and clone. Link to github documentation on the "Github Flow".

Now you are ready to contribute!

## Running Tests Locally

First, [set up your development environment](#setting-up-a-development-environment).

> Our tests are managed by `tox`, which runs both `pytest` tests and our system of makefiles.

Author: ```suggestion
Our tests are managed by `tox`, which runs both `pytest` tests and our system of makefiles.
All `tox` commands *must* be issued from the project root directory.
```

> The regression does not end on the first failure, but continues until all tests in the `test` and `example` directories have been run.

Author: ```suggestion
The regression does not end on the first failure, but continues until all tests in the `/tests` and `/examples` directories have been run.
```

To run the tests locally with `tox`, you will need to select an appropriate test environment.
Valid test environments are formatted as `{your python version}-{your OS}`.
Valid python version values are `py35`, `py36`, `py37`, `py38`, or `py39`;
and valid OS values are `linux`, `macos`, or `windows`.
For example, a valid test environment is `py38-linux`.
You can see the list of valid test environments by running the below command.

```command
tox -l
```

Once you know the test environment you wish to use, call `tox` .

```command
tox -e py38-linux
```

At the end of the regression, if there were any test failures, the tests that failed will be printed.
Otherwise, tox will print a green `:)`.

> ### Selecting a Regression Language and Simulator

Author: ```suggestion
### Selecting a Language and Simulator for Regression
```

> `tox` supports the usage of the environment variables `SIM` and `TOPLEVEL_LANG` to select a simulator and language to run the regression.

Author: ```suggestion
`tox` supports the usage of the environment variables [`SIM`](https://docs.cocotb.org/en/stable/building.html#var-SIM) and [`TOPLEVEL_LANG`](https://docs.cocotb.org/en/stable/building.html#var-TOPLEVEL_LANG) to select a simulator and language to run the regression.
```

By default the tests will attempt to run with the Icarus Verilog simulator.

> For example, if you wanted to run tests with GHDL on Linux with Python 3.8, you would issue the following command.

Author: ```suggestion

For example, if you wanted to run tests with GHDL on Linux with Python 3.8, you would issue the following command.
```

```command
SIM=ghdl TOPLEVEL_LANG=vhdl tox -e py38-linux
```

### Running Individual Tests Locally

> Each test under `/tests/test_cases/*/` and `/examples/tests/` can be run individually.

Author: ```suggestion
Each test under `/tests/test_cases/*/` and `/examples/*/tests/` can be run individually.
```

This is particularly useful if you want to run a particular test that fails the regression.

> First you must install cocotb from source by navigating to the project root and issuing the following.

Author: ```suggestion
First you must install cocotb from source by navigating to the project root directory and issuing the following.
```

```command
python -m pip install .
```

> On Windows, you must install cocotb from source like so.

Author: ```suggestion
On Windows, you must instead install cocotb from source like so.
```

```command
python -m pip install --global-option build_ext --global-option --compiler=mingw32 .
```

Once that has been done, you can navigate to the directory containing the test you wish to run.

> Then you may issue an [appropriate](https://docs.cocotb.org/en/stable/quickstart.html#running-your-first-example) `make` command.

Author: ```suggestion
Then you may issue an [appropriate]https://docs.cocotb.org/en/stable/building.html#makefile-based-test-scripts) `make` command.
```

```command
make SIM=icarus
```

## Building Documentation Locally

First, [set up your development environment](#setting-up-a-development-environment).

Documentation is built locally using `tox`.
The last message in the output will contain a URL to the documentation you just built.
Simply copy and paste the link into your browser to view it.
The documentation will be built in the same location on your hard drive on every run, so you only have to refresh the page to see new changes.

> To build the documentation locally on Linux or Mac, issue the following command.

Author: ```suggestion
To build the documentation locally on Linux or Mac, issue the following command from the project root directory.
```

First, [set up your development environment](#setting-up-a-development-environment).

Our tests are managed by `tox`, which runs both `pytest` tests and our system of makefiles.

> The regression does not end on the first failure, but continues until all tests in the `/tests` and `/examples` directories have been run.

Reviewer 2: If tests fail the examples are not run (https://github.com/cocotb/cocotb/runs/1301183158?check_suite_focus=true)

Author: Yeah that's a bug that's been around a while. I'd rather not report that as a fact.

Reviewer: We also don't (or at least used to) run all examples in regression, but I don't think that level of details needs to be mentioned here.
